### PR TITLE
Added dataset parser to cli.py

### PR DESCRIPTION
**Pull request by DESm1th**
---
name: Added a dataset cli parser
about: Added a crawler to make use of the dats model files
title: 'Feature implementation | Dataset Crawler'
labels: enhancement
assignees: ''

---

## Related issues

Related to: https://github.com/CONP-PCNO/conp-portal/issues/76

## Checklist

- [x] **DO** Unit tests pass.
- [ ] **DO** If new feature, created unit test.

## Purpose

This adds a 'update_datasets' cli option that will scrape the data sets in the portal's app/static/data/projects folder and either add a new record to the database (if the dataset ID and annex uuid are new) or find and update an existing record.

## Current behaviour

Currently the database 'datasets' table cannot be populated based on the datasets folder.

## New behaviour

The datasets table of the database can be populated based on the contents of app/static/data/projects by running `flask update_datasets`
 
#### Does this introduce a major change?
- [ ] Yes
- [x] No

## Implementation Detail

There's a new function in app/cli.py. I added a few support functions as protected functions in that script because I wasnt sure if there was somewhere better to put them. Also, a few of the database columns I wasnt sure how to fill in: 'image', 'owner_id'  (users table not yet populated) and 'category'.

Also, (in spite of my best efforts) the code got a bit unruly. I think that it may be hard to adequately pull info out the dats.json files without normalizing a few of the columns in the current database schema. It might also make adding new datasets to the database easier if we stored the dats.json files as a 'jsonb' column in the database instead of trying to extract values, since we'd get the entire file contents if anything else in there ends up being useful later on,  but I think this might restrict the portal to using postgresql (I'm not sure which other rdbms support the json / jsonb column type).

I did not add any new unit tests but I can. I just wanted to get feedback to see if I went in the right direction first :)

### CODE REVIEW

File: app/cli.py [hunk 1]
```diff
@@ -63,9 +70,146 @@ def seed_test_db():
...
+                name = None
+                modality = None
+                version = None
+                data_format = None
+                is_private = None
+
+            record = Dataset(dataset_id=dataset_id,
+                             annex_uuid=annex_uuid,
+                             description=description,
+                             owner_id=None,
```
Comment (shots47s) [inline app/cli.py]: So owner is not in the DATS? right?

File: app/cli.py [hunk 2]
```diff
@@ -63,9 +70,146 @@ def seed_test_db():
...
+                             download_path=download_path,
+                             raw_data_url=raw_data_url,
+                             image=None,
+                             name=name,
+                             modality=modality,
+                             version=version,
+                             format=data_format,
+                             category=None,
```
Comment (shots47s) [inline app/cli.py]: Same with category, maybe we should consider changing this to "Keywords"

File: app/cli.py [hunk 3]
```diff
@@ -7,9 +7,16 @@
 import os
 import click
 import csv
+import configparser
+import json
+import logging
```
Comment (shots47s) [inline app/cli.py]: Adding logging.... I like it!

File: app/cli.py (hunk 1 above)
Comment (DESm1th) [reply app/cli.py]: There's a field that looks right, but the current database schema expects an integer so I assumed it was meant to link to an entry in the user table (there's no foreign key constraint on that column though so I could be wrong)

File: app/cli.py [hunk 4]
```diff
@@ -63,9 +70,146 @@ def seed_test_db():
...
 
+    @app.cli.command("update_datasets")
+    def update_datasets():
+        from app import db
+        from app.models import Dataset
+        from sqlalchemy.exc import IntegrityError
+
+        logger.setLevel(app.logger.getEffectiveLevel())
+
+        data_path = app.config.get("DATA_PATH", "static/data/projects").lstrip("/")
```
[the remark this answers is not in the document]
Comment (shots47s) [reply app/cli.py]: Actually, we should make this so it relies on the env variable more and defines only the relative path within the app.  So I would change this to something like:

`search_path = Path(app.config.get("DATA_PATH")).joinpath("projects")`

File: app/cli.py [hunk 5]
```diff
@@ -63,9 +70,146 @@ def seed_test_db():
...
+                private = model.get("privacy") == "Open Access"
+            else:
+                description = None
+                download_path = None
+                raw_data_url = None
+                name = None
+                modality = None
+                version = None
+                data_format = None
+                is_private = None
```
Comment (shots47s) [inline app/cli.py]: is_private is not defined outside of here, so if it doesn't hit this else, it doesn't know it.

**Comment by shots47s** [comment]
Should make it so that if the data is not present for the dataset, no data is inserted into the database.  Right now, if a dataset fails to read a dats.json file, it gets an empty entry in the database with the annex ID.  Check to see if dats is there before committing to the database to keep the database pure.

File: app/cli.py [hunk 6]
```diff
@@ -63,9 +70,146 @@ def seed_test_db():
...
+        value = parser[section][key]
+    except Exception as e:
+        logger.info("Couldnt parse {}.{} from {}. Reason - {} {}".format(
+                section, key, file_path, type(e).__name__, e))
+        return None
+    return value
+
+
+def _get_dats_model(project):
+    dats_file = project.joinpath("DATS.json")
```
[the remark this answers is not in the document]
Comment (xlecours) [reply app/cli.py]: FYI https://github.com/conpdatasets/1000GenomesProject/pull/1
One offender fixed. Any suggestion on how to assert that the dats.json exists and is lowercase?

**Comment by liamocn** [comment]
I'm inclined to go ahead and merge this if nobody's got any problems? There's a conflict that needs resolving but I don't think it's anything major.

**Comment by shots47s** [comment]
I would hold off, what Xavier is working on will supersede this.

**Comment by liamocn** [comment]
> 
> 
> I would hold off, what Xavier is working on will supersede this.

OK, will do

**Comment by xlecours** [comment]
Following enhancement on the DATS.json standardization, a DATSDataset class was created to access the descriptor properties that can be in various formats.
Most of the 'ambiguous' columns have been removed from the Dataset table. 
The datasets are now installed and updated in app/cli.py as was proposed in this PR. But it is now using datalad to install the conp-dataset and subdatasets recursively. 

Closing this.

**Pull request closed without merging**